shell/platform/android/image_external_texture_gl.cc

```diff
@@ -64,12 +62,35 @@ bool ImageExternalTextureGL::MaybeSwapImages() {
   JavaLocalRef old_android_image(android_image_);
   android_image_.Reset(image);
   JavaLocalRef hardware_buffer = HardwareBufferFor(image);
-  egl_image_ = CreateEGLImage(AHardwareBufferFor(hardware_buffer));
+  AHardwareBuffer* latest_hardware_buffer = AHardwareBufferFor(hardware_buffer);
+  auto key = flutter::NDKHelpers::AHardwareBuffer_getId(latest_hardware_buffer);
+  auto existing_image = FindImage(key);
+  if (existing_image != nullptr) {
+    dl_image_ = existing_image;
+
+    CloseHardwareBuffer(hardware_buffer);
+    // IMPORTANT: We have just received a new frame to display so close the
+    // previous Java Image so that it is recycled and used for a future frame.
+    CloseImage(old_android_image);
+    return;
+  }
+
+  egl_image_ = CreateEGLImage(latest_hardware_buffer);
```

Comment:
I don't think that the handling of egl_image_ is correct anymore. We have the following dependency:

DlImage -> eglImage -> HardwareBuffer.

And we have a cache from HardareBuffer -> DlImage.

I think the correct thing to do is to include the eglImage in the cache and only destroy the eglImage when evicting from the cache.

As it stands now every time we create a new DlImage we end up running the destructor for the previous frame's egl_image_ which should make the cache'd DlImage point at a destroyed image?

I'm kind of surprised this works, maybe we don't need to keep the eglImage alive for as long as the DlImage is alive for? If that is the case, we should still remove egl_image_ as a field and make it a local variable here and pass it into CreateDlImage (which currently sniffs the egl_image_ field).

Reply:
Maybe the code is working today because we never have a cache hit?

Reply:
I tried making the egl image lifespan essentially just for the scope of this method as suggested. Things seem to work correctly.

One thing I have noticed is that at least on the current test phone (a Pixel 4), Impeller GLES or Vulkan is much smoother than Skia. I'm not sure if this due to a bug in how the caching is handled (I need to try with this patch disabled) or just the performance characteristics of the phone.